src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs

```diff
@@ -545,26 +545,34 @@ public ImmutableArray<T> InsertRange(int index, IEnumerable<T> items)
             Requires.NotNull(items, "items");
 
             if (self.Length == 0)
-            {
                 return ImmutableArray.CreateRange(items);
-            }
-
-            int count = ImmutableExtensions.GetCount(ref items);
-            if (count == 0)
+            
+            if (items is ICollection<T>)
             {
-                return self;
+                ICollection<T> other = (ICollection<T>)items;
+                
+                if (other.Count == 0)
+                    return self;
+                
+                T[] dest = new T[self.Length + other.Count];
+                other.CopyTo(dest, index); // keep this before so any changes it makes to unintended parts of the array are overwritten by Array.Copy
+                Array.Copy(self.array, 0, dest, 0, index);
+                Array.Copy(self.array, index, dest, index + other.Count, self.Length - index);
+                return new ImmutableArray<T>(dest);
             }
-
-            T[] tmp = new T[self.Length + count];
-            Array.Copy(self.array, 0, tmp, 0, index);
-            int sequenceIndex = index;
-            foreach (var item in items)
+            else
             {
-                tmp[sequenceIndex++] = item;
+                T[] other = items.ToArray();
+                
+                if (other.Length == 0)
+                    return self;
+    
+                T[] tmp = new T[self.Length + other.Length];
+                Array.Copy(self.array, 0, tmp, 0, index);
+                Array.Copy(other, 0, tmp, index, other.Length);
+                Array.Copy(self.array, index, tmp, index + other.Length, self.Length - index);
+                return new ImmutableArray<T>(tmp);
```

Comment:
This change impacts both behavior and performance, and not necessarily in a positive way.  For example, the previous implementation was using GetCount, which is able to get a count from an ```ICollection```, an ```ICollection<T>```, or an ```IReadOnlyCollection<T>```, whereas the new version is only specializing ```ICollection<T>```.  There's too much changing as part of this for me to feel confident that it's a net benefit.  Do you have performance tests that demonstrate it is across a range of input collections and collection types?  If so, can you share?  I suggest this be reverted for now.